sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/OfferProcessor.java

```diff
@@ -0,0 +1,319 @@
+                    offers.stream()
+                            .map(offer -> offer.getId())
+                            .collect(Collectors.toList()));
+
+            LOGGER.info("Enqueuing {} offer{}. Updated offers in progress: {}",
+                    offers.size(),
+                    offers.size() == 1 ? "" : "s",
+                    offersInProgress.stream()
+                            .map(offerID -> offerID.getValue())
+                            .collect(Collectors.toList()));
+        }
+
+        for (Protos.Offer offer : offers) {
+            boolean queued = offerQueue.offer(offer);
+            if (!queued) {
+                LOGGER.warn("Offer queue is full: Declining offer and removing from in progress: '{}'",
+                        offer.getId().getValue());
+                declineShort(Arrays.asList(offer));
+                // Remove AFTER decline: Avoid race where we haven't declined yet but appear to be done
+                synchronized (inProgressLock) {
+                    offersInProgress.remove(offer.getId());
+                }
+            }
+        }
+
+        if (!multithreaded) {
+            // Process on this thread, rather than depending on offerExecutor to do it.
+            processQueuedOffers();
+        }
+    }
+
+    public void dequeue(Protos.OfferID offerId) {
+        offerQueue.remove(offerId);
+    }
+
+    /**
+     * All offers must have been presented to resourceOffers() before calling this.  This call will block until all
+     * offers have been processed.
+     *
+     * @throws InterruptedException if waiting for offers to be processed is interrupted
+     * @throws IllegalStateException if offers were not processed in a reasonable amount of time
+     */
+    @VisibleForTesting
+    public void awaitOffersProcessed() throws InterruptedException {
+        final int totalDurationMs = 5000;
+        final int sleepDurationMs = 100;
+        for (int i = 0; i < totalDurationMs / sleepDurationMs; ++i) {
+            synchronized (inProgressLock) {
+                if (offersInProgress.isEmpty()) {
+                    LOGGER.info("All offers processed.");
+                    return;
+                }
+                LOGGER.warn("Offers in progress {} is non-empty, sleeping for {}ms ...",
+                        offersInProgress.stream().map(id -> id.getValue()).collect(Collectors.toList()),
+                        sleepDurationMs);
+            }
+            Thread.sleep(sleepDurationMs);
+        }
+        throw new IllegalStateException(String.format(
+                "Timed out after %dms waiting for offers to be processed", totalDurationMs));
+    }
+
+    /**
+     * Dequeues and processes any elements which are present on the offer queue, potentially blocking for offers to
+     * appear.
+     */
+    private void processQueuedOffers() {
+        LOGGER.info("Waiting for queued offers...");
+        List<Protos.Offer> offers = offerQueue.takeAll();
+        try {
```

Comment:
Not clear to me why this try block exists.  Doesn't mean it shouldn't.

Reply:
It's mainly for the following `finally`, to ensure we increment the offers processed in the metrics, even if an exception is thrown